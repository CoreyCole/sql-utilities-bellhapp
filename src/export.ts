import * as shelljs from 'shelljs';
import * as prettyJSONStringify from 'pretty-json-stringify';
import * as moment from 'moment';

import { queries } from './queries';

const rejectHandler = (err) => {
  console.error(err);
  throw err;
};

export namespace exportScripts {
  /**
   * Queries all of the data for a restaurant location given a restaurantName
   * @param {mysql.Connection} connection
   * @param {string} restaurantName
   */
  export async function getRestaurantLocation (connection, restaurantName) {
    const exportTime = moment().format('YYYY-MM-DD-HH-mm-ss');
    const restaurantLocationRow = await queries.getRestaurantLocation(connection, restaurantName).catch(rejectHandler);
    const restaurantLocation = { _exportDate: exportTime, ...restaurantLocationRow };
    restaurantLocation.menus = {};
    const menuRows = await queries.getMenus(connection, restaurantLocation.id).catch(rejectHandler);
    for (const menuRow of menuRows) {
      restaurantLocation.menus[menuRow.uid] = { ...menuRow };
      restaurantLocation.menus[menuRow.uid].sections = {};
      const sectionRows = await queries.getMenuSections(connection, menuRow.id).catch(rejectHandler);
      for (const sectionRow of sectionRows) {
        restaurantLocation.menus[menuRow.uid].sections[sectionRow.uid] = { ...sectionRow };
        restaurantLocation.menus[menuRow.uid].sections[sectionRow.uid].items = {};
        const itemRows = await queries.getMenuSectionItems(connection, sectionRow.id).catch(rejectHandler);
        for (const itemRow of itemRows) {
          restaurantLocation.menus[menuRow.uid].sections[sectionRow.uid].items[itemRow.uid] = { ...itemRow };
          restaurantLocation.menus[menuRow.uid].sections[sectionRow.uid].items[itemRow.uid].optionGroups = {};
          const optionGroupRows = await queries.getItemOptionGroups(connection, itemRow.id).catch(rejectHandler);
          if (optionGroupRows) {
            for (const optionGroupRow of optionGroupRows) {
              restaurantLocation.menus[menuRow.uid].sections[sectionRow.uid].items[itemRow.uid].optionGroups[optionGroupRow.uid] = { ...optionGroupRow };
              restaurantLocation.menus[menuRow.uid].sections[sectionRow.uid].items[itemRow.uid].optionGroups[optionGroupRow.uid].options = {};
              const optionGroupOptionRows = await queries.getItemOptionGroupOptions(connection, optionGroupRow.id).catch(rejectHandler);
              for (const optionGroupOptionRow of optionGroupOptionRows) {
                restaurantLocation.menus[menuRow.uid].sections[sectionRow.uid].items[itemRow.uid].optionGroups[optionGroupRow.uid].options[optionGroupOptionRow.uid] = { ...optionGroupOptionRow };
              }
            }
          }
        }
      }
    }
    return restaurantLocation;
  }
  /**
   * Exports the provided json in "pretty" format to the passed file (appends export date onto file name)
   * @param {object} restaurantLocation
   * @param {string} exportFileName
   */
  export function writeJsonToFile (restaurantLocation, exportPath) {
    shelljs.echo(prettyJSONStringify(restaurantLocation, {
      tab: '  ',
      spaceBeforeColon: ''
    })).toEnd(`${exportPath}/export-${restaurantLocation['_exportDate']}.json`);
  }
  /**
   * Returns a markdown string representing the passed restaurant location
   * @param restaurantLocation the json representation of the restaurant location
   */
  export function convertJsonToMarkdown (restaurantLocation): string {
    let markdown: string[] = [];
    markdown.push(`<h1 style='margin:0;'>${restaurantLocation.restaurantName} Menu</h1>`);
    for (const menuUid in restaurantLocation.menus) {
      const menu = restaurantLocation.menus[menuUid];
      markdown.push(`<h2 style='margin:0;margin-top:16px;margin-bottom:-32px;'>${menu.menuName}</h2>`);
      for (const sectionUid in menu.sections) {
        const section = menu.sections[sectionUid];
        markdown.push(`<h3 style='margin:0;margin-top:16px;margin-bottom:-16px;'>${section.sectionName}</h3>`);
        for (const itemUid in section.items) {
          const item = section.items[itemUid];
          markdown.push(`$${(item.price / 100).toFixed(2)} - ${item.itemName}`);
          for (const optionGroupUid in item.optionGroups) {
            const optionGroup = item.optionGroups[optionGroupUid];
            markdown.push(`* ${optionGroup.type} - ${optionGroup.optionGroupName}`);
            for (const optionUid in optionGroup.options) {
              const option = optionGroup.options[optionUid];
              markdown.push(`  * $${(option.value / 100).toFixed(2)} - ${option.optionGroupOptionName}`);
            }
            markdown.push('\n');
          }
        }
      }
    }
    return markdown.join('\n');
  }
}
