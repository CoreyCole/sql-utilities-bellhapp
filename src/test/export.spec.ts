import * as mysql from 'mysql';

import { exportScripts } from '../export';

const RESTAURANT_NAME = 'Cedars';
const CONFIG = require('../config.json');
const CONNECTION = mysql.createConnection(CONFIG);

describe('json export', () => {
  // WARNING: this takes super long to run, run `npm run test-watch`
  //          so that this only gets tested once if you're not changing
  //          related files
  it('should export restaurant location data given a restaurant name', async () => {
    jasmine.DEFAULT_TIMEOUT_INTERVAL = 15000;
    const restaurantLocation = await exportScripts.getRestaurantLocation(CONNECTION, RESTAURANT_NAME);
    expect(restaurantLocation).toBeDefined();
    expect(restaurantLocation.id).toBeDefined();
    expect(restaurantLocation.uid).toBeDefined();
    expect(restaurantLocation.restaurantName).toBeDefined();
    expect(restaurantLocation.menus).toBeDefined();
  });
});
