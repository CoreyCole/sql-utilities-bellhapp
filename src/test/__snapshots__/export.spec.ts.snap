// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pdf convert should export a restaurant's json to a markdown string 1`] = `
"<h1 style='margin:0;'>Cedars Menu</h1>
<h2 style='margin:0;margin-top:16px;margin-bottom:-32px;'>Appetizers</h2>
<h3 style='margin:0;margin-top:16px;margin-bottom:-16px;'>Appetizers</h3>
$4.50 - Hummus
* OPTION_ADD - Sides
  * $2.95 - Chai

"
`;
